internal/blocksync/reactor.go

```diff
@@ -489,94 +485,112 @@ func (r *Reactor) poolRoutine(ctx context.Context, stateSynced bool, blockSyncCh
+				newBlockPartSetHeader = newBlockParts.Header()
+				newBlockID            = types.BlockID{Hash: newBlock.Hash(), PartSetHeader: newBlockPartSetHeader}
 			)
 
-			// Finally, verify the first block using the second's commit.
-			//
-			// NOTE: We can probably make this more efficient, but note that calling
-			// first.Hash() doesn't verify the tx contents, so MakePartSet() is
-			// currently necessary.
-			if err = state.Validators.VerifyCommitLight(chainID, firstID, first.Height, second.LastCommit); err != nil {
-				err = fmt.Errorf("invalid last commit: %w", err)
-				r.logger.Error(
-					err.Error(),
-					"last_commit", second.LastCommit,
-					"block_id", firstID,
-					"height", first.Height,
-				)
+			if r.lastTrustedBlock != nil {
+				err := VerifyNextBlock(newBlock, newBlockID, verifyBlock, r.lastTrustedBlock.block, r.lastTrustedBlock.commit, state.NextValidators)
+
+				if err != nil {
+					switch err.(type) {
+					case ErrBlockIDDiff:
+					case ErrValidationFailed:
+						peerID := r.pool.RedoRequest(r.lastTrustedBlock.block.Height + 1)
+						if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
+							NodeID: peerID,
+							Err:    errors.New("invalid block for verification"),
+						}); serr != nil {
+							return
+						}
+					case ErrInvalidVerifyBlock:
+						r.logger.Error(
+							err.Error(),
+							"last_commit", verifyBlock.LastCommit,
+							"block_id", newBlockID,
+							"height", r.lastTrustedBlock.block.Height,
+						)
+						peerID := r.pool.RedoRequest(r.lastTrustedBlock.block.Height + 2)
+						if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
+							NodeID: peerID,
+							Err:    err,
+						}); serr != nil {
+							return
+						}
+					}
 
-				// NOTE: We've already removed the peer's request, but we still need
-				// to clean up the rest.
-				peerID := r.pool.RedoRequest(first.Height)
-				if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
-					NodeID: peerID,
-					Err:    err,
-				}); serr != nil {
-					return
 				}
 
-				peerID2 := r.pool.RedoRequest(second.Height)
-				if peerID2 != peerID {
-					if serr := blockSyncCh.SendError(ctx, p2p.PeerError{
-						NodeID: peerID2,
-						Err:    err,
-					}); serr != nil {
-						return
+			} else {
+				// we need to load the last block we trusted
+				if r.initialState.LastBlockHeight != 0 {
+					r.lastTrustedBlock = &BlockResponse{r.store.LoadBlock(r.initialState.LastBlockHeight), r.store.LoadBlockCommit(r.initialState.LastBlockHeight)}
+					if r.lastTrustedBlock == nil {
+						panic("Failed to load last trusted block")
 					}
 				}
-			} else {
-				r.pool.PopRequest()
-
-				// TODO: batch saves so we do not persist to disk every block
-				r.store.SaveBlock(first, firstParts, second.LastCommit)
+				oldHash := r.initialState.Validators.Hash()
+				if !bytes.Equal(oldHash, newBlock.ValidatorsHash) {
 
-				var err error
+					fmt.Println(
```

Comment:
Do you mean `r.logger.Error(` ?

Reply:
Also, please add a meaningful message before the variables

Reply:
Changed this to a valid error and added a request redo here.